lib/sample_rate_converter/sample_rate_converter.c

```diff
@@ -0,0 +1,407 @@
+
+	} else {
+		arm_err = arm_fir_decimate_init_q15(&ctx->fir_decimate_q15, filter_size,
+						    ctx->conversion_ratio, (q15_t *)filter_coeffs,
+						    ctx->state_buf_15,
+						    CONFIG_SAMPLE_RATE_CONVERTER_BLOCK_SIZE);
+	}
+#elif CONFIG_SAMPLE_RATE_CONVERTER_BIT_DEPTH_32
+	if (ctx->conversion_direction == CONVERSION_DIR_UP) {
+		arm_err = arm_fir_interpolate_init_q31(&ctx->fir_interpolate_q31,
+						       ctx->conversion_ratio, filter_size,
+						       (q31_t *)filter_coeffs, ctx->state_buf_31,
+						       CONFIG_SAMPLE_RATE_CONVERTER_BLOCK_SIZE);
+	} else {
+		arm_err = arm_fir_decimate_init_q31(&ctx->fir_decimate_q31, filter_size,
+						    ctx->conversion_ratio, (q31_t *)filter_coeffs,
+						    ctx->state_buf_31,
+						    CONFIG_SAMPLE_RATE_CONVERTER_BLOCK_SIZE);
+	}
+#endif
+	if (arm_err == ARM_MATH_LENGTH_ERROR) {
+		LOG_ERR("Filter size is not a multiple of conversion ratio");
+		return -EINVAL;
+	} else if (arm_err != ARM_MATH_SUCCESS) {
+		LOG_ERR("Unknown error during interpolator/decimator initialization (%d)", arm_err);
+		return -EINVAL;
+	}
+
+	LOG_DBG("Sample rate converter initialized. Input sample rate: %d, Output sample rate: %d, "
+		"conversion ratio: %d, filter type: %d",
+		ctx->input_sample_rate, ctx->output_sample_rate, ctx->conversion_ratio,
+		ctx->filter_type);
+	return 0;
+}
+
+int sample_rate_converter_open(struct sample_rate_converter_ctx *ctx)
+{
+	if (ctx == NULL) {
+		LOG_ERR("Context cannot be NULL");
+		return -EINVAL;
+	}
+
+	memset(ctx, 0, sizeof(struct sample_rate_converter_ctx));
+
+	return 0;
+}
+
+int sample_rate_converter_process(struct sample_rate_converter_ctx *ctx,
+				  enum sample_rate_converter_filter filter, void *input,
+				  size_t input_size, uint32_t input_sample_rate, void *output,
+				  size_t output_size, size_t *output_written,
+				  uint32_t output_sample_rate)
+{
+	int ret;
+	uint8_t *read_ptr;
+	uint8_t *write_ptr;
+	size_t samples_to_process;
+
+	uint8_t internal_input_buf[SAMPLE_RATE_CONVERTER_INTERNAL_INPUT_BUF_SIZE];
+	uint8_t internal_output_buf[SAMLPE_RATE_CONVERTER_INTERNAL_OUTPUT_BUF_SIZE];
+
+#if CONFIG_SAMPLE_RATE_CONVERTER_BIT_DEPTH_16
+	size_t bytes_per_sample = sizeof(uint16_t);
+#elif CONFIG_SAMPLE_RATE_CONVERTER_BIT_DEPTH_32
+	size_t bytes_per_sample = sizeof(uint32_t);
+#endif
+
+	if (input_size % bytes_per_sample != 0) {
+		LOG_ERR("Size of input is not a byte multiple");
+		return -EINVAL;
+	}
+
+	size_t samples_in = input_size / bytes_per_sample;
+
+	if (samples_in > CONFIG_SAMPLE_RATE_CONVERTER_BLOCK_SIZE) {
+		LOG_ERR("Too many samples given as input");
+		return -EINVAL;
+	}
+
+	if ((ctx == NULL) || (input == NULL) || (output == NULL) || (output_written == NULL)) {
+		LOG_ERR("Null pointer received");
+		return -EINVAL;
+	}
+
+	if ((ctx->input_sample_rate != input_sample_rate) ||
+	    (ctx->output_sample_rate != output_sample_rate) || (ctx->filter_type != filter)) {
+		LOG_DBG("State has changed, re-initializing filter");
```

Comment:
Do we expect this to change run-time? Perhaps this should be a WRN?

Reply:
My understanding is that the audio application allows the stream to change sample rate during run time, and I wrote it like this so the application does not need to care if the sample rate changes. If this is not the case we should definitely discuss changing it! 👍 

As for the log message I think it makes the most sense to have debug as it is a completely valid case the way the lib is written now.

Reply:
The conversion can absolutely change run-time. If you are going from a phone call to high quality music for instance, or even medium quality (24kHz)